language/resource-api/README.md

```diff
@@ -333,33 +344,101 @@ When a resource is marked with `noop => true`, either locally or through a globa
+Declaring this feature restricts the resource from being run "locally". It is expected to execute all external interactions through the `context.transport` instance. The way that instance is set up is runtime specific. In Puppet, it is configured through the [`device.conf`](https://puppet.com/docs/puppet/5.3/config_file_device.html) file, and only available when running under [`puppet device`](https://puppet.com/docs/puppet/5.3/man/device.html).
+
+### Transport
+
+```ruby
+# lib/puppet/transport/nexus_schema.rb
+Puppet::ResourceAPI.register_transport(
+  name: 'nexus', # points at class Puppet::Transport::Nexus
+  desc: 'Connects to a Cisco Nexus device'
+  features: [], # TODO
+  connection_info: {
+    connector_type: {
+      type: 'Enum[nxapi, ssh, grpc, pidgeon]',
+      desc: 'Which method to use to connect to the host'
+    },
+    hostname: {
+      type: 'String',
+      desc: 'The host to connect to.',
+    },
+    username: {
+      type: 'String',
+      desc: 'The user.',
+    },
+    password: {
+      type: 'Sensitive[String]',
+      desc: 'The password to connect.',
+    },
+    enable_password: {
+      type: 'Sensitive[String]',
+      desc: 'The password escalate to enable access.',
+    },
+    port: {
+      type: 'Integer',
+      desc: 'The port to connect to.',
+    },
+  },
+)
+
+#  lib/puppet/transport/nexus.rb
+class Puppet::Transport::Nexus
+  def initialize(transport_info); end
+  def facts; end
+  def verify; end
+  def close; end
+end
+```
+
+The transport layer consists of the schema and the implementation. The schema is defined in the same manner as a `Type`, except instead of `attributes`,
+ there is now `connection_info` and it describes the shape of the data which should be passed to the implementation for a connection to be made.
+
+The implementation must follow the structure as defined above and provide the following methods:
+
+  * initialize(connection_info)
+    * `transport_info` contains validated hash matching schema
+  * facts
+    * Access the device and return a facts hash
+  * verify
+    * Perform an operation that returns true if the connection is correctly established
+  * close
+    * Close the connection
+
+### Legacy Support
+
+Prior support for `remote_resource` was provided through `Puppet::Util::NetworkDevice`. This code can be updated by inheriting from `Puppet::ResourceApi::Transport::Forwarder` rather than `Puppet::Util::NetworkDevice::Simple::Device`.
```

Comment:
Is there a reason we don't just make `Puppet::Util::NetworkDevice::Simple::Device` the forwarder?

Reply:
ah cool this makes more sense.